Source/JavaScriptCore/wasm/WasmStruct.cpp

```diff
@@ -0,0 +1,154 @@
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WasmStruct.h"
+
+#if ENABLE(WEBASSEMBLY)
+
+#include "JSCJSValueInlines.h"
+
+namespace JSC::Wasm {
+
+constexpr size_t typeKindSize(TypeKind kind)
+{
+    switch (kind) {
+    case TypeKind::I32:
+    case TypeKind::F32: {
+        return 32;
+    }
+    case TypeKind::I64:
+    case TypeKind::F64: {
+        return 64;
+    }
+
+    case TypeKind::Funcref:
+    case TypeKind::Externref:
+    case TypeKind::Ref:
+    case TypeKind::RefNull: {
+        return sizeof(WriteBarrierBase<Unknown>);
+    }
+    case TypeKind::Func:
+    case TypeKind::Struct:
+    case TypeKind::Void:
+    case TypeKind::Rec:
+    case TypeKind::I31ref: {
+        break;
+    }
+    }
+
+    ASSERT_NOT_REACHED();
+    return 0;
+}
+
+size_t WasmStruct::structDataAllocationSize(const StructType* structType)
+{
+    size_t totalSize = 0u;
+    for (StructFieldCount i = 0; i < structType->fieldCount(); ++i)
+        totalSize += typeKindSize(structType->field(i).type.kind);
+    return totalSize;
+}
+
+WasmStruct::WasmStruct(Ref<TypeDefinition>&& type)
+    : m_type(WTFMove(type))
+{
+}
+
+uint64_t WasmStruct::get(uint32_t fieldIndex) const
+{
+    ASSERT(fieldIndex < structType().fieldCount());
+    const uint32_t* targetPointer = reinterpret_cast<const uint32_t*>(reinterpret_cast<const uint8_t*>(this) + offsetOfTail() + getFieldOffset(fieldIndex));
+
+    switch (fieldKind(fieldIndex)) {
+    case TypeKind::I32:
+    case TypeKind::F32:
+        return *targetPointer;
+    case TypeKind::I64:
+        return *reinterpret_cast<const uint64_t*>(targetPointer);
+    case TypeKind::F64:
+        return *reinterpret_cast<const uint64_t*>(targetPointer);
+    case TypeKind::Externref:
+    case TypeKind::Funcref:
+    case TypeKind::Ref:
+    case TypeKind::RefNull:
+        return JSValue::encode(bitwise_cast<WriteBarrierBase<Unknown>*>(targetPointer)->get());
+    default:
+        ASSERT_NOT_REACHED();
+        return 0;
+    }
+}
+
+size_t WasmStruct::getFieldOffset(uint32_t fieldIndex) const
+{
+    ASSERT(fieldIndex < structType().fieldCount());
+
+    size_t fieldOffset = 0;
+    for (unsigned i = 0; i < fieldIndex; ++i)
+        fieldOffset += typeKindSize(fieldKind(i));
+    return fieldOffset;
+}
+
+void WasmStruct::set(JSGlobalObject* globalObject, uint32_t fieldIndex, JSValue argument)
+{
+    uint8_t* targetPointer = reinterpret_cast<uint8_t*>(this) + offsetOfTail() + getFieldOffset(fieldIndex);
```

Comment:
nit: bitwise_cast is a little nicer here

Reply:
agree, fixed